listenbrainz/webserver/views/playlist_api.py

```diff
@@ -0,0 +1,516 @@
+                                           APIInternalServerError,
+                                           APINotFound,
+                                           APIForbidden)
+from listenbrainz.webserver.rate_limiter import ratelimit
+from listenbrainz.webserver.views.api import _validate_auth_header
+from listenbrainz.webserver.views.api_tools import log_raise_400, is_valid_uuid
+from listenbrainz.db.model.playlist import Playlist, WritablePlaylist, WritablePlaylistRecording
+
+playlist_api_bp = Blueprint('playlist_api_v1', __name__)
+
+PLAYLIST_TRACK_URI_PREFIX = "https://musicbrainz.org/recording/"
+PLAYLIST_ARTIST_URI_PREFIX = "https://musicbrainz.org/artist/"
+PLAYLIST_RELEASE_URI_PREFIX = "https://musicbrainz.org/release/"
+PLAYLIST_URI_PREFIX = "https://listenbrainz.org/playlist/"
+PLAYLIST_EXTENSION_URI = "https://musicbrainz.org/doc/jspf#playlist"
+PLAYLIST_TRACK_EXTENSION_URI = "https://musicbrainz.org/doc/jspf#track"
+RECORDING_LOOKUP_SERVER_URL = "https://labs.api.listenbrainz.org/recording-mbid-lookup/json"
+MAX_RECORDINGS_PER_ADD = 100
+
+
+def _parse_boolean_arg(name, default=None):
+    value = request.args.get(name)
+    if not value:
+        return default
+
+    value = value.lower()
+    if value not in ["true", "false"]:
+        raise APIBadRequest("Invalid %s argument: %s. Must be 'true' or 'false'" % (name, value))
+
+    return True if value == "true" else False
+
+
+def validate_playlist(jspf, require_title=True):
+    """
+        Given a JSPF dict, ensure that title is present and that if tracks are present
+        they have valid URIs or MBIDs specified. If any errors are found 400 is raised.
+
+        If a created_for field is found and the user is not an approved playlist bot,
+        then a 403 forbidden will be raised.
+    """
+
+    if require_title:
+        try:
+            title = jspf["playlist"]["title"]
+            if not title:
+                raise KeyError
+        except KeyError:
+            log_raise_400("JSPF playlist must contain a title element with the title of the playlist.")
+
+    if 'track' not in jspf:
+        return
+
+    for i, track in enumerate(jspf['track']):
+        try:
+            recording_uri = track["identifier"]
+            if not recording_uri:
+                raise KeyError
+        except KeyError:
+            log_raise_400("JSPF playlist track %d must contain an identifier element with recording MBID." % i)
+
+        if recording_uri.startswith(PLAYLIST_TRACK_URI_PREFIX):
+            recording_mbid = recording_uri[len(PLAYLIST_TRACK_URI_PREFIX):]
+        else:
+            log_raise_400("JSPF playlist track %d identifier must have the namespace '%s' prepended to it." %
+                          (i, PLAYLIST_TRACK_URI_PREFIX))
+
+        if not is_valid_uuid(recording_mbid):
+            log_raise_400("JSPF playlist track %d does not contain a valid track identifier field." % i)
+
+
+def serialize_jspf(playlist: Playlist):
```

Comment:
consider adding this method to the `Playlist` class instead, so that we can just `aplaylist.as_jspf()`

Reply:
sounds sane, but the not duplicating code between troi and LB is still open. Let's resolve that first, then this will fall out...